Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+StateMachine.swift

```diff
@@ -0,0 +1,222 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+extension HTTPConnectionPool {
+    struct StateMachine {
+        struct Action {
+            let request: RequestAction
+            let connection: ConnectionAction
+
+            init(request: RequestAction, connection: ConnectionAction) {
+                self.request = request
+                self.connection = connection
+            }
+
+            static let none: Action = Action(request: .none, connection: .none)
+        }
+
+        enum ConnectionAction {
+            enum IsShutdown: Equatable {
+                case yes(unclean: Bool)
+                case no
+            }
+
+            case createConnection(Connection.ID, on: EventLoop)
+            case scheduleBackoffTimer(Connection.ID, backoff: TimeAmount, on: EventLoop)
+
+            case scheduleTimeoutTimer(Connection.ID, on: EventLoop)
+            case cancelTimeoutTimer(Connection.ID)
+
+            case closeConnection(Connection, isShutdown: IsShutdown)
+            case cleanupConnections(CleanupContext, isShutdown: IsShutdown)
+
+            case none
+        }
+
+        enum RequestAction {
+            case executeRequest(Request, Connection, cancelTimeout: Bool)
+            case executeRequestsAndCancelTimeouts([Request], Connection)
+
+            case failRequest(Request, Error, cancelTimeout: Bool)
+            case failRequestsAndCancelTimeouts([Request], Error)
+
+            case scheduleRequestTimeout(for: Request, on: EventLoop)
+            case cancelRequestTimeout(Request.ID)
+
+            case none
+        }
+
+        enum HTTPTypeStateMachine {
+            case http1(HTTP1StateMachine)
+        }
+
+        var state: HTTPTypeStateMachine
```

Comment:
Assuming `HTTPTypeStateMachine` will only have `http1` and `http2` cases (at least for now!) then `httpVersion`/`httpType` or similar might be a clearer name here since we're really just delegating to the state machine each case holds.